### ISSUE TITLE
chef-client, in daemon mode, when called with a HUP doesn’t release old logging file handles

### ISSUE THREAD

**Issue by cachamber** [chef/chef#3435]
**Issue:**  When chef-client is run in daemon mode, with log rotation, it holds onto all logging file handles after a rotation, as it never closes the old log.
**Reproduction:**
1. Start chef-client in daemon mode
2. Validate log file handle with lsof 
3. Move the logfile to another name to simulate log rotation
4. Validate log file handle has moved to the new filename
5. Use init script to reload chef-client (Sends a HUP to the process)
6. Verify, using lsof that chef-client now has 2 log file handles
```
[root@xldcac01 ~]# chef-client --version
Chef: 12.3.0
[root@xldcac01 ~]# ps -efa |grep chef-client 
root     25450     1  0 13:29 ?        00:00:00 /opt/chef/embedded/bin/ruby /usr/bin/chef-client -d -c /etc/chef/client.rb -L /var/log/chef/client.log -P /var/run/chef/client.pid -i 1800 -s 20 -f
[root@xldcac01 ~]# lsof -p 25450 |grep client.log
chef-clie 25450 root   10w   REG  253,2      244     1875 /var/log/chef/client.log
[root@xldcac01 ~]# mv /var/log/chef/client.log /var/log/chef/client.log-1
[root@xldcac01 ~]# lsof -p 25450 |grep client.log
chef-clie 25450 root   10w   REG  253,2      244     1875 /var/log/chef/client.log-1
[root@xldcac01 ~]# /etc/init.d/chef-client reload
Reloading chef-client:                                     [  OK  ]
[root@xldcac01 ~]# lsof -p 25450 |grep client.log
chef-clie 25450 root   10w   REG  253,2      309     1875 /var/log/chef/client.log-1
chef-clie 25450 root   12w   REG  253,2        0     1858 /var/log/chef/client.log
[root@xldcac01 ~]# 
```
**Location:** Found in https://github.com/chef/chef/blob/master/lib/chef/application.rb
**Analysis:**
The HUP handler is on lines 77-80, calls “reconfigure” on line 79.
“reconfigure” defined on lines 47-53  the issue is that reconfigure calls “configure_logging” on line 49
“configure_logging” is defined on line 137, the offending line is line 138: 
```
Chef::Log.init(MonoLogger.new(Chef::Config[:log_location]))
```
There isn't a check to see if a logger already exists.
**Fix:**
Before creating a new instance of the logger, check to see if one exists, and close it.
```
# Check to see if a logger already exists, if so close it
if Chef::Log.info?
     Chef::Log.close
end
```

**Comment by yozik04** [comment]
A year has passed and still no fix...

**Comment by fzbishopclark** [comment]
> A year has passed and still no fix...
Actually, the fix is available under mhorbul's patch on 2016-03-18 .  It's just gotta be merged in if it's not there yet.  It's so close!

**Comment by thommay** [comment]
Just merged chef/mixlib-log#13 to mixlib-log that should resolve this.
